### PR TITLE
Lock files attempt 2

### LINKED ISSUE #4288: Write pantry hashes into a .lock file instead of emitting warnings
Currently `stack build` will emit warnings if I do not specify `size` and `sha256` for `github` dependencies in `stack.yaml`.

Running `stack freeze` will output the relevant information that I then can add to `stack.yaml` manually.

I propose to write the information that is currently produced by `stack freeze` to a `.lock` file after every successful build (e.g. `pantry.lock`):

- when the `.lock` file exists then for each listed dependency:
  - `stack` uses the information as if it were directly listed in `stack.yaml`; specifically it errors out on any hash mismatches
  - when `stack` detects that a dependency was updated in `stack.yaml` it silently ignores the information from the `.lock` file
- after a successful build `stack` creates the `.lock` file, overwriting any existing versions (which in effect means that we remove unused entries and add missing entries)

Regarding revision control, we can [follow conventions that have been developed elsewhere](https://yehudakatz.com/2010/12/16/clarifying-the-roles-of-the-gemspec-and-gemfile/):
When developing an application, the `.lock` file should be checked into revision control.  When developing a library, the `.lock` file should be added to `.gitignore`.

A word on naming: If we expect the `.lock` file to be specific to `stack` we may want to name it `stack.lock` (or possibly `.stack.lock` to not interfere with file name completion); if on the other hand we think that it will be relevant for other build tools that implement support for `pantry` then we could name it `pantry.lock`.

Assuming we adopt this approach we can:
- remove any warnings related to pantry hashes
- remove the `stack freeze` command

CC @snoyberg

### PR DESCRIPTION
**Pull request by qrilka**
Please include the following checklist in your PR:

* [ ] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [ ] The documentation has been updated, if necessary.

Tested manually and with local run of integration tests
Has some code and resolves #4550 
Resolves #4288

### CODE REVIEW

**Comment by qrilka** [comment]
One note about the implementation: it persist all dependency locations and that means that `stack.yaml.lock` is quite large because it contains all snapshot packages so loading it takes noticeable time.

**Comment by snoyberg** [comment]
I haven't looked at the code yet, but based on your comment @qrilka: do you think it makes more sense to take a different approach here? Two things I can think of:

* Don't include all packages. This seems like it will make it quite difficult to come up with a `SourceMapHash`
* Use a more efficient format, either as the primary format or as an optimized format for lookup. In the latter, I mean keeping the .lock file as a YAML file, but keeping a .lock.bin file next to it (or elsewhere) in a more efficient binary format, or maybe even storing in the SQLite database

**Comment by qrilka** [comment]
@snoyberg I think I need to check the current timings to get a bit more clear understanding of the problem. But in any case it looks odd to add an optimization which makes Stack slower.

**Comment by qrilka** [comment]
@snoyberg on Monday I've tried to optimize sourcemap-based lockfiles and I had some improvements but still if we parse almost the same what we have in a snapshot it takes some time. So in 40bf6f0 I've simplified everything and in that commit lock file contains only non-trivial location completions (e.g. snapshot locations won't get stored there) - it simplifies logic quite a lot and also it gives some speed improvements (up to 20% compared to optimized sourcemap-based locks).
I think we should go this simpler and a bit faster way.
One visible difference with the previous version is that we don't check snapshots but I don't see how that could cause any troubles.
If we choose this simple variant then `lock_files.md` needs to be rewritten.

**Comment by qrilka** [comment]
BTW regarding timings I have this gist - https://gist.github.com/qrilka/d5e220f3955c8d8ea70ad40f720f5f12

**Comment by qrilka** [comment]
@snoyberg 1 of integration test failed, so I need to fix that before it will be ready for a proper review

**Comment by qrilka** [comment]
the problem with `1438-configure-options` looks to be unrelated to this change so this PR is ready to be reviewed

File: src/test/Stack/LockSpec.hs [hunk 1]
```diff
@@ -0,0 +1,182 @@
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Stack.LockSpec where
```
Comment (snoyberg) [inline src/test/Stack/LockSpec.hs]: Very happy to see this

File: subs/pantry/src/Pantry.hs [hunk 2]
```diff
@@ -396,7 +399,7 @@ loadCabalFileImmutable
 loadCabalFileImmutable loc = withCache $ do
   logDebug $ "Parsing cabal file for " <> display loc
   bs <- loadCabalFileBytes loc
-  let foundCabalKey = BlobKey (SHA256.hashBytes bs) (FileSize (fromIntegral (B.length bs)))
+  let foundCabalKey = bsToBlobKey bs
```
Comment (snoyberg) [inline subs/pantry/src/Pantry.hs]: Good helper function

File: src/test/Stack/LockSpec.hs (hunk 1 above)
Comment (qrilka) [reply src/test/Stack/LockSpec.hs]: Thanks to @psibi :)

**Comment by qrilka** [comment]
@snoyberg please take another look, especially into the doc

File: doc/lock_files.md [hunk 3]
```diff
@@ -206,36 +177,8 @@ If the lock file does not exist, it will be created by:
...
-3. For each incomplete package location:
-    * Lookup the value in the map created in (1)
-    * If present: use that completed information
-    * Otherwise: complete the information using the same completion
-      procedure from Pantry as in "creation"
-
-This should minimize the number of changes to packages incurred.
+When loading a Stack project all completed package or snapshot locations
+(even when they were completed using information from a lock file) get
+collected to form a new lock file.
```
Comment (snoyberg) [inline doc/lock_files.md]: > to form a new lock file

Can you clarify this a bit? I'm guessing the logic is "form a new lock file in memory and compare against the one on disk, writing if there are any differences." Is that the case?

Comment (qrilka) [reply doc/lock_files.md]: Yes, I'll add this

Comment (qrilka) [reply doc/lock_files.md]: done

**Pull request merged**